### ISSUE TITLE
Semantic tokens: No documentation on how/why to use resultId

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#86166]
Testing #85976

resultId/previousResultId appears in several APIs but there aren't any docs around it so I couldn't test it.

**Comment by Tyriar** [comment]
> *NOTE*: It is illegal to return `SemanticTokensEdits` if `options.previousResultId` is not set!

@alexandrudima I was under the impression an extension would listen to document changes and update using those events, that would mean it could provide `SemanticTokensEdits` all the time?

**Comment by alexandrudima** [comment]
@Tyriar No, the contract is that the semantic tokens are "pulled", not "pushed". (I go a bit in depth [here](https://github.com/microsoft/vscode/issues/86168#issuecomment-561538432)). An easier mental model for me is a client-server situation:

* initially, the client makes the request to get semantic tokens with `previousResultId: undefined` and no ranges.
* the server receives this requests and understands that it must return all the tokens in the file (in full)
* the server returns a result, generates a result id and stores the result in memory.
* ... the user types in the file repeatedly ... these edits make it to the server via already existing document change mechanisms ...
* the client eventually makes the request to get semantic tokens and passes in the `previousResultId`
* the server computes the new semantic tokens and checks if it still has in memory / if it still remembers `previousResultId`. If that is the case, and if the delta would be smaller than sending the results in full, the server **may** return a delta. If the server has dropped the previous result so it cannot diff the new results with the old results, it can send the new results in full.

**Comment by dbaeumer** [comment]
@alexandrudima the following doesn't read nicely:

> it is safe and smart for a smart tokens provider can compute a diff from the previous result to the new result.

**Comment by alexandrudima** [comment]
@dbaeumer Thanks, I have rewritten the JSDoc... pushed to master.
